modules/afsocket/afsocket-grammar.ym

```diff
@@ -884,9 +884,30 @@ inet_socket_option
 	| KW_IP_TOS '(' nonnegative_integer ')'               { ((SocketOptionsInet *) last_sock_options)->ip_tos = $3; }
 	| KW_IP_FREEBIND '(' yesno ')'              { ((SocketOptionsInet *) last_sock_options)->ip_freebind = $3; }
 	| KW_INTERFACE '(' string ')'		    { socket_options_inet_set_interface_name((SocketOptionsInet *) last_sock_options, $3); free($3); }
-	| KW_TCP_KEEPALIVE_TIME '(' nonnegative_integer ')'   { ((SocketOptionsInet *) last_sock_options)->tcp_keepalive_time = $3; }
-	| KW_TCP_KEEPALIVE_INTVL '(' nonnegative_integer ')'  { ((SocketOptionsInet *) last_sock_options)->tcp_keepalive_intvl = $3; }
-	| KW_TCP_KEEPALIVE_PROBES '(' nonnegative_integer ')' { ((SocketOptionsInet *) last_sock_options)->tcp_keepalive_probes = $3; }
+	| KW_TCP_KEEPALIVE_TIME '(' nonnegative_integer ')'
+	   {
+#ifdef TCP_KEEPIDLE
```

Comment:
My only issue with this solution is that we a referencing those macros redundantly, once in the implementation, once in the grammar, opening the possibility that it would get inconsistent again.

What if we created a HAVE_TCP_KEEPALIVE_TIMERS feature flag, defined by configure.ac that could be used in both places. This feature flag would be set by configure if all the required socket options would be available.

With that we could even swap the 3 distinct conditionals with a single one in both places.

Reply:
@bazsi : I have defined HAVE_TCP_KEEPALIVE_TIMERS in configure.ac now